dbt/exceptions.py

```diff
@@ -101,8 +102,62 @@ def type(self):
         return 'Compilation'
 
 
+EXTRA_PROPERTIES = re.compile(
+    r'Additional properties are not allowed \(([^)]*) (were|was) unexpected\)'
+)
+MISSING_PROPERTIES = re.compile(r'(.*) is a required property')
+
+
+def _extract_error_cause_types(errors):
+    """Given an array of errors, generate three lists:
+        - list of extra values that must be removed
+        - list of missing values that must be added
+        - list of errors that aren't either of those
+    """
+    extras = []
+    missing = []
+    unknown = []
+    for err in errors:
+        match = EXTRA_PROPERTIES.match(err)
+        if match:
+            extras.extend(match.group(1).split(', '))
+            continue
+        match = MISSING_PROPERTIES.match(err)
+        if match:
+            missing.append(match.group(1))
+            continue
+        unknown.append(err)
+    return extras, missing, unknown
+
+
+def _causes_from_errors(errors):
+    """Generate an error content causes string from the array of errors returned
+    by the validators.
+    """
+    extras, missing, unknown = _extract_error_cause_types(errors)
+    parts = []
+    if extras:
+        parts.append(
+            "Extra project configuration value(s) {} not recognized"
+            .format(', '.join(extras)))
+    if missing:
+        parts.append(
+            "Project configuration value(s) {} not supplied"
+            .format(', '.join(missing)))
+    if unknown:
+        parts.extend(("Unrecognized error: '{}'".format(e) for e in unknown))
+    if not parts:
+        parts = ["Unable to read credentials"]
+    return ' and '.join(parts)
+
+
 class ValidationException(RuntimeException):
-    pass
+    def __init__(self, validator_name, errors, node=None):
+        self.validator_name = validator_name
+        self.causes = _causes_from_errors(errors)
+        msg = ('Invalid arguments passed to "{}" instance: {}'
+               ).format(validator_name, self.causes)
+        super(ValidationException, self).__init__(msg, node)
```

Comment:
i like the idea of accepting structured data here. have you checked everywhere in the code that uses ValidationException to make sure this is compatible?

this ValidationException is generic across all of dbt -- it applies to other things, like models and schema.yml, etc. so i don't think this is the right place to wire in _causes_from_errors that refers specifically to project configuration.